classes/local/store/s3/client.php

```diff
@@ -430,10 +500,104 @@ public function generate_presigned_url($contenthash, $headers) {
+     *
+     * @return string.
+     */
+    private function generate_presigned_cloudfront_url($contenthash = '', $headers = array(), $nicefilename = true) {
+
+        if (!$this->enablepresignedcloudfronturls) {
+            return ''; // Throw an exception - called but not enabled.
+        }
+
+        $cdnconfig = get_objectfs_config();
+        $key = $this->get_filepath_from_hash($contenthash);
+        $cloudfrontclient = new CloudFrontClient(
+            array(
+                'profile' => 'default',
+                'version' => 'latest', /* '2014-11-06' */
+                'region' => $cdnconfig->s3_region,  /* The region is the source bucket region ? - 'ap-southeast-2' */
+            )
+        );
+
+        $resourcedomain = $cdnconfig->cloudfront_resource_domain;
+
+        $resourcekey = $resourcedomain . '/' . $key;
+
+        if ($nicefilename) {
+            // We are trying to deliver original filename rather than hash filename to client.
+
+            $contentdisposition = '';
+            $originalfilename = '';
+            $originalcontenttype = '';
+
+            $contentdisposition = $this->get_header($headers, 'Content-Disposition');
+            if ($contentdisposition !== '') {
+                $contentdisposition = trim($contentdisposition); // S3 $params['ResponseContentDisposition'].
+            }
+
+            $contenttype = $this->get_header($headers, 'Content-Type');
+            if ($contenttype !== '') {
+                $originalcontenttype = trim($contenttype); // S3 $params['ResponseContentType'].
+            }
+
+            /*
+                Need to get the filename and content-type from HEADERS array
+                Without invoking more DB hits (the header array contains it already by now).
+            */
+
+            if (!empty($contentdisposition)) {
+                $fparts = explode('; ', $contentdisposition);
+                $originalfilename = str_replace('filename=', '', $fparts[1]); // Get the actual filename.
+                $originalfilename = str_replace('"', '', $originalfilename); // Remove the quotes.
+                $contentdisposition = $fparts[0];
+
+                $newkey = $key .
+                    '?response-content-disposition='.rawurlencode(
+                        $contentdisposition . ';filename="' . utf8_encode($originalfilename) . '"'
+                    ) .
+                    '&response-content-type=' . rawurlencode($originalcontenttype);
+
+                // Alternative without filename: $newkey = $key . '?response-content-disposition='.rawurlencode($contentdisposition.';'.(utf8_encode($originalfilename)).'').'&response-content-type='.rawurlencode($originalcontenttype);.
+
+                $resourcekey = $resourcedomain . '/' . $newkey;
+            }
+        }
+
+        if (isset($cdnconfig->expirationtime) && !empty($cdnconfig->expirationtime)) {
+            $expirationvalue = time() + $cdnconfig->expirationtime;
```

Comment:
this feels wrong, on line 585 we will end up with:

 time() + time() + $cdnconfig->expirationtime

Reply:
I'd ditch $expirationvalue and just move $expires up in the if else